### PR TITLE
Rewriting to use pure python power

### PR DESCRIPTION
**Pull request by hamishivi**
This PR removes all the pyspark stuff and replaces it with some pure python goodness. I've used a library called [implicit](https://github.com/benfred/implicit), which seems up to date, and also seems to give similar results to before.

Please give this branch a go and let me know if there are any issues, or fix it yourself! I didn't want to commit to master cause I haven't thoroughly tested this.

I also didn't really change the layout/way the site works- but damn, it'd be so good to get async stuff working cause then we could just add the user's profile to the data, and give recs in a matter of a second or two, instead of waiting for the friend graph search to finish.

### CODE REVIEW

File: app/rec.py [hunk 1]
```diff
@@ -1,64 +1,48 @@
...
-    return sc.parallelize(data_list).map(lambda r: Rating(int_hash(r[0]), int(r[1]), int(r[2])))
-
-def int_hash(s):
-    return hash(s) % 2147483647
+    # lets weight these models by bm25weight. see https://en.wikipedia.org/wiki/Okapi_BM25
+    # this essentially weights values by their relevance (ie if a game is played an unusually
+    # large amount of time compared to most other people)
+    # So this should ideally help stop the Valve bias - if everyone plays a lot of DOTA2, then it should
+    # have a lower weight.
+    plays = implicit.nearest_neighbours.bm25_weight(gameplay_time, K1=100, B=0.8)
```
Comment (hamishivi) [inline app/rec.py:24]: [this](http://opensourceconnections.com/blog/2015/10/16/bm25-the-next-generation-of-lucene-relevation/) article has an intuitive look at this, which I found really interesting personally.

File: app/views.py [hunk 2]
```diff
@@ -7,49 +7,60 @@
...
+    # load from the file.
+    global data, game_matrix
+    # try to load
+    try:
+        print('loading training data')
+        data, game_matrix = rec.load('./training_data')
+    # if something goes wrong, trace a random user and then try again!
+    except Exception:
+        print('no data found, loading random user data')
+        u_info = user_info.get_user_data(76561198045011271)
```
Comment (KittenHero) [inline app/views.py:21]: https://www.xkcd.com/221/

File: app/views.py [hunk 3]
```diff
@@ -7,49 +7,60 @@
...
-    print('user' in g)
     if 'user' in session and session['user'] is not None:
         unplayed_games = user_info.get_unplayed_games(session['user'])
-        recs = rec.get_rec(int(session['user']), 10000)
-        #print(recs)
-        naive = False
-        if recs is None and 'naive' not in session:
+        data, game_matrix = rec.load('./training_data')
+        naive =  False
+        # we have already trawled this user, just give them their recs!
```
Comment (KittenHero) [inline app/views.py:39]: trawling?

Comment (hamishivi) [reply app/views.py:39]: as in looked at them and their friends (trawl = search thoroughly)

File: app/views.py (hunk 2 above)
Comment (hamishivi) [reply app/views.py:21]: ;)

**Pull request merged**